File: apps/web/lib/activity.ts
import { createServiceClient } from '@/lib/supabase/server'

interface LogActivityParams {
  userId: string
  eventType: string
  urlId?: string
  scanId?: string
  payload?: Record<string, unknown>
}

/**
 * Append one row to `activity_log`. Uses the service-role client because the
 * table has no client INSERT policy (RLS). Best-effort: a logging failure must
 * never change the API response that triggered it.
 */
export async function logActivity(params: LogActivityParams): Promise<void> {
  try {
    const supabase = createServiceClient()
    await supabase.from('activity_log').insert({
      user_id: params.userId,
      event_type: params.eventType,
      url_id: params.urlId ?? null,
      scan_id: params.scanId ?? null,
      payload: params.payload ?? {},
    })
  } catch {
    // best-effort — swallow
  }
}

File: apps/scanner/lib/activity.py
from lib.supabase import get_supabase


def log_event(
    user_id: str,
    event_type: str,
    *,
    url_id: str | None = None,
    scan_id: str | None = None,
    payload: dict | None = None,
) -> None:
    """Append one row to `activity_log` (service-role client — the table has
    no client INSERT policy). Best-effort: logging is observability, not the
    job, so a failure here is swallowed and never propagates."""
    try:
        get_supabase().table("activity_log").insert({
            "user_id": user_id,
            "event_type": event_type,
            "url_id": url_id,
            "scan_id": scan_id,
            "payload": payload or {},
        }).execute()
    except Exception:
        pass
